src/v3/mod.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{
+    aes::{self, aes_encrypt_with_iv},
+    icl_header_v3::{v3document_header::Header, V3DocumentHeader},
+    signing::AES_KEY_LEN,
+};
+use bytes::Bytes;
+use protobuf::Message;
+
+const IV_LEN: usize = 12;
+const GCM_TAG_LEN: usize = 16;
+
+struct V3Signature {
+    iv: [u8; IV_LEN],
+    gcm_tag: [u8; GCM_TAG_LEN],
+}
+
+fn decompose_signature(sig: &Bytes) -> Option<V3Signature> {
+    if sig.len() < IV_LEN + GCM_TAG_LEN {
+        None
+    } else {
+        let (iv, _) = sig.split_at(IV_LEN);
+        let (_, gcm_tag) = sig.split_at(sig.len() - GCM_TAG_LEN);
+        Some(V3Signature {
+            iv: iv.try_into().unwrap(),           // Length was validated up-front
+            gcm_tag: gcm_tag.try_into().unwrap(), // Length was validated up-front
+        })
+    }
+}
+
+pub fn verify_signature(key: [u8; AES_KEY_LEN], v3_header: &V3DocumentHeader) -> bool {
+    // If we have no header or authTag, that means this document was encrypted before the header was added, which would only happen if the
+    // document was encrypted with the Java SDK. In that case, we'll just ignore the verification and try to decrypt, which might still work.
+    if v3_header.header.is_none() || !v3_header.has_saas_shield() || v3_header.sig.is_empty() {
+        true
+    } else {
+        let maybe_sig = decompose_signature(&v3_header.sig);
+        match (maybe_sig, &v3_header.header) {
+            (Some(sig), Some(Header::SaasShield(header))) => aes_encrypt_with_iv(
+                aes::EncryptionKey(key),
+                &header
+                    .write_to_bytes()
+                    .expect("Writing proto to bytes failed."),
+                sig.iv,
+                &[],
+            )
+            .map(|(_, new_sig)| {
+                let new_sig_length = new_sig.0.len();
+                let (_, new_gcm_tag) = new_sig.0.split_at(new_sig_length - GCM_TAG_LEN);
+                new_gcm_tag == sig.gcm_tag
+            })
+            .unwrap_or(false),
+            _ => false,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I'd like to see a test with with no signature and an invalid signature as well since that will exercise the failure conditions.